### PR TITLE
Mqtt: Fix NPE and other issues

### PR DESCRIPTION
**Pull request by davidgraeff**
Now that I start to use the io.transport.mqtt bundle to implement some of the ideas in #3812, I encountered some problems that this PR addresses.

## NPE
The ReconnectPolicy family of classes need an internal reference to the MqttBrokerConnection (to call start() again). The reference is set when `MqttBrokerConnection.setReconnectPolicy` is called. That was at least the plan, but didn't happen. The result was an NPE in the PeriodicReconnectPolicy object.

## Tests
* Move test classes from the mqtt.test package to the mqtt package to have access to package private methods.
* Add extensive tests for ReconnectPolicy (there were none before).
* Extend tests for MqttConnectionObserver.

## Necessary API changes
* Rename MqttConnectionObserver.setConnected to MqttConnectionObserver.connectionStateChanged and add a second Throwable parameter to gain access to the error that happened during a connect. There was no way to retrieve any error message for displaying it to the user before.
* Make addConnectionObserver not immediately call the callback back. This is unnecessary and breaks a simple pattern to call `MqttBrokerConnection.start()` synchronously (more details in the _java doc_ for `start()`).
* Add own MqttException instead of leaking the internal paho MqttException.
* Add own MqttPublishCallback and MqttPublishResult to not leak the internal paho callback interface.

As usual there are some added JavaDocs and comments here and there when I touched the ReconnectPolicy and related classes.

### CODE REVIEW

File: bundles/io/org.eclipse.smarthome.io.transport.mqtt/src/main/java/org/eclipse/smarthome/io/transport/mqtt/MqttConnectionObserver.java [hunk 1]
```diff
@@ -11,14 +11,18 @@
...
+     * requires that the control is returned quickly to not stall the Mqtt thread.
      *
-     * @param connected true if the connection is established, false if disconnected.
+     * @param connected True if the connection is established, false if disconnected.
+     * @param error An exception object (might be a MqttException) with the reason why
+     *            a connection failed.
      */
-    public void setConnected(boolean connected);
-
+    public void connectionStateChanged(boolean connected, Throwable error);
```
Comment (htreu) [inline bundles/io/org.eclipse.smarthome.io.transport.mqtt/src/main/java/org/eclipse/smarthome/io/transport/mqtt/MqttConnectionObserver.java]: This is an API breaking change. If we get aways with this, how about "extending" it a little further by passing an enum `ConnectionState` with `CONNECTED, DISCONNECTED` instead of the boolean. This will allow us to extend the API w/o breaking it. wdyt?

File: bundles/io/org.eclipse.smarthome.io.transport.mqtt/src/main/java/org/eclipse/smarthome/io/transport/mqtt/reconnect/AbstractReconnectPolicy.java [hunk 2]
```diff
@@ -22,10 +22,21 @@
...
     }
 
+    /**
+     * Return the brokerConnection object that this reconnect policy is assigned to.
+     */
+    public MqttBrokerConnection getBrokerConnection() {
+        return brokerConnection;
+    }
+
+    public boolean isReconnecting() {
```
Comment (htreu) [inline bundles/io/org.eclipse.smarthome.io.transport.mqtt/src/main/java/org/eclipse/smarthome/io/transport/mqtt/reconnect/AbstractReconnectPolicy.java]: Is this also a candidate for being `abstract`? This would force the implementations to think about what to return here.

File: bundles/io/org.eclipse.smarthome.io.transport.mqtt/src/main/java/org/eclipse/smarthome/io/transport/mqtt/MqttConnectionObserver.java (hunk 1 above)
Comment (davidgraeff) [inline bundles/io/org.eclipse.smarthome.io.transport.mqtt/src/main/java/org/eclipse/smarthome/io/transport/mqtt/MqttConnectionObserver.java]: I agree and have it implemented this way. The API is only 1 day old, that's why I think if we act 'quick', it is worth it. Before actually using the API in a project I haven't noticed that there is no way of retrieving any information on what went wrong for a connection attempt.

**Comment by davidgraeff** [comment]
@htreu: I just noticed, that the manifest file was incomplete and some classes are in the wrong package. Could you please remove your approval and do a review of the second commit as well. This is what happened:
* Export org.eclipse.smarthome.io.transport.mqtt.reconnect
* Export org.eclipse.smarthome.io.transport.mqtt.sslcontext
* Move classes from org.eclipse.smarthome.io.transport.sslcontext to org.eclipse.smarthome.io.transport.**mqtt**.sslcontext
* Because nobody was able to use the org.eclipse.smarthome.io.transport.mqtt.reconnect namespace anyway so far, this should not be an API breakage: Rename Reconnect**Policy** to Reconnect**Strategy**. Strategy is the more correct term here.
* Add constructors to `PeriodicReconnectStrategy`.

**Comment by davidgraeff** [comment]
And I've got this question: Some methods in `MqttBrokerConnection` can throw a `MqttException`. This is a `org.eclipse.paho.client.mqttv3` class and forces users of the `MqttBrokerConnection` class to import that package. Is this all right or should we encapsulate this exception in an own `MqttBrokerConnection` to hide the implementation details?

**Comment by sjka** [comment]
> Is this all right or should we encapsulate this exception in an own `MqttException` to hide the implementation details?

Please do the latter to keep the API clean and not leak the library classes.

**Comment by davidgraeff** [comment]
In a third commit, I have added a `MqttException` class to no longer leak the internal library one.

File: bundles/io/org.eclipse.smarthome.io.transport.mqtt/src/main/java/org/eclipse/smarthome/io/transport/mqtt/MqttWillAndTestament.java [hunk 3]
```diff
@@ -66,6 +66,28 @@ public static MqttWillAndTestament fromString(String string) {
...
+     * @param payload The optional payload. Can be null.
+     * @param qos Valid values are 0 (Deliver at most once),1 (Deliver at least once) or 2</li>
+     * @param retain true if messages shall be retained
+     */
+    public MqttWillAndTestament(String topic, byte[] payload, int qos, boolean retain) {
+        this.topic = topic;
+        this.payload = payload;
+        this.qos = qos;
+        this.retain = retain;
+        if (StringUtils.isBlank(this.topic)) {
```
Comment (htreu) [inline bundles/io/org.eclipse.smarthome.io.transport.mqtt/src/main/java/org/eclipse/smarthome/io/transport/mqtt/MqttWillAndTestament.java]: this should be the first call/block in the constructor.

File: bundles/io/org.eclipse.smarthome.io.transport.mqtt/src/main/java/org/eclipse/smarthome/io/transport/mqtt/MqttWillAndTestament.java [hunk 4]
```diff
@@ -66,6 +66,28 @@ public static MqttWillAndTestament fromString(String string) {
         return result.isValid() ? result : null;
     }
 
+    private MqttWillAndTestament() {
+
```
Comment (htreu) [inline bundles/io/org.eclipse.smarthome.io.transport.mqtt/src/main/java/org/eclipse/smarthome/io/transport/mqtt/MqttWillAndTestament.java]: please comment the empty block to indicate why its empty.

File: bundles/io/org.eclipse.smarthome.io.transport.mqtt/src/main/java/org/eclipse/smarthome/io/transport/mqtt/MqttException.java [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+    private Throwable cause;
+
+    /**
+     * Constructs a new <code>MqttException</code> with the specified code
+     * as the underlying reason.
+     *
+     * @param reasonCode the reason code for the exception.
+     */
+    public MqttException(int reasonCode) {
+        super();
```
Comment (davidgraeff) [inline bundles/io/org.eclipse.smarthome.io.transport.mqtt/src/main/java/org/eclipse/smarthome/io/transport/mqtt/MqttException.java]: Are u sure? Because the base classes (Exception, Throwable) do this as well and in throwable there is actually work done. I thought it is common to do this in exceptions. I will remove this if you are absolutely sure though.

File: bundles/io/org.eclipse.smarthome.io.transport.mqtt/src/main/java/org/eclipse/smarthome/io/transport/mqtt/MqttWillAndTestament.java (hunk 3 above)
Comment (davidgraeff) [inline bundles/io/org.eclipse.smarthome.io.transport.mqtt/src/main/java/org/eclipse/smarthome/io/transport/mqtt/MqttWillAndTestament.java]: done

File: bundles/io/org.eclipse.smarthome.io.transport.mqtt/src/main/java/org/eclipse/smarthome/io/transport/mqtt/MqttWillAndTestament.java (hunk 4 above)
Comment (davidgraeff) [inline bundles/io/org.eclipse.smarthome.io.transport.mqtt/src/main/java/org/eclipse/smarthome/io/transport/mqtt/MqttWillAndTestament.java]: done

**Comment by davidgraeff** [comment]
I discovered, that the publish() method leaked another paho class (IMqttActionListener) and therefore added MqttPublishCallback and MqttPublishResult. The MqttBrokerConnection.publish() method now also returns the messageID.

File: bundles/io/org.eclipse.smarthome.io.transport.mqtt/src/main/java/org/eclipse/smarthome/io/transport/mqtt/MqttException.java (hunk 5 above)
Comment (htreu) [inline bundles/io/org.eclipse.smarthome.io.transport.mqtt/src/main/java/org/eclipse/smarthome/io/transport/mqtt/MqttException.java]: Please see the last `Note` in the official Java documentation: https://docs.oracle.com/javase/tutorial/java/IandI/super.html

The no-arg super constructor is always called.

Comment (davidgraeff) [inline bundles/io/org.eclipse.smarthome.io.transport.mqtt/src/main/java/org/eclipse/smarthome/io/transport/mqtt/MqttException.java]: done

**Comment by davidgraeff** [comment]
@htreu I have incorporated all suggested changes

**Comment by kaikreuzer** [comment]
Thanks @davidgraeff!
